src/phoenix/server/api/dataloaders/token_counts.py

```diff
@@ -0,0 +1,110 @@
+from typing import (
+    Any,
+    AsyncContextManager,
+    Callable,
+    DefaultDict,
+    List,
+    Literal,
+    Optional,
+    Tuple,
+)
+
+from openinference.semconv.trace import SpanAttributes
+from sqlalchemy import Select, func, select
+from sqlalchemy.ext.asyncio import AsyncSession
+from sqlalchemy.sql.functions import coalesce
+from strawberry.dataloader import DataLoader
+from typing_extensions import TypeAlias
+
+from phoenix.db import models
+from phoenix.server.api.input_types.TimeRange import TimeRange
+from phoenix.trace.dsl import SpanFilter
+
+Kind: TypeAlias = Literal["prompt", "completion", "total"]
+ProjectRowId: TypeAlias = int
+TimeInterval: TypeAlias = Tuple[Optional[datetime], Optional[datetime]]
+FilterCondition: TypeAlias = Optional[str]
+TokenCount: TypeAlias = int
+
+Segment: TypeAlias = Tuple[TimeInterval, FilterCondition]
+Param: TypeAlias = Tuple[ProjectRowId, Kind]
+
+Key: TypeAlias = Tuple[Kind, ProjectRowId, Optional[TimeRange], FilterCondition]
+Result: TypeAlias = TokenCount
+ResultPosition: TypeAlias = int
+DEFAULT_VALUE = 0
+
+
+class TokenCountDataLoader(DataLoader[Key, Result]):
+    def __init__(self, db: Callable[[], AsyncContextManager[AsyncSession]]) -> None:
+        super().__init__(load_fn=self._load_fn, cache_key_fn=self._cache_key_fn)
+        self._db = db
+
+    @staticmethod
+    def _cache_key_fn(key: Key) -> Tuple[Segment, Param]:
+        kind, project_rowid, time_range, filter_condition = key
+        interval = (
+            (time_range.start, time_range.end)
+            if isinstance(time_range, TimeRange)
+            else (None, None)
+        )
+        return (interval, filter_condition), (project_rowid, kind)
+
+    async def _load_fn(self, keys: List[Key]) -> List[Result]:
+        results: List[Result] = [DEFAULT_VALUE] * len(keys)
+        arguments: DefaultDict[
+            Segment,
+            DefaultDict[Param, List[ResultPosition]],
+        ] = defaultdict(lambda: defaultdict(list))
+        for position, key in enumerate(keys):
+            segment, param = self._cache_key_fn(key)
+            arguments[segment][param].append(position)
+        async with self._db() as session:
+            for segment, params in arguments.items():
+                stmt = _get_stmt(segment, *params.keys())
+                data = await session.stream(stmt)
+                async for project_rowid, prompt, completion, total in data:
+                    for position in params[(project_rowid, "prompt")]:
+                        results[position] = prompt
+                    for position in params[(project_rowid, "completion")]:
+                        results[position] = completion
+                    for position in params[(project_rowid, "total")]:
+                        results[position] = total
+        return results
+
+
+def _get_stmt(
+    segment: Segment,
+    *params: Param,
+) -> Select[Any]:
+    (start_time, end_time), filter_condition = segment
```

Comment:
I think it would help for readability to unpack these in the function signature.

Reply:
The idea here is to maintain a semblance of regularity by repeating the `Segment` and `Param` monikers. By restructuring it upfront, the regularity is lost.